Add performJITMemcpyAtomic and simplify jit copying code.

justinmichaud opened the pull request:
#### 06f449154f24e7747855d1c1101d97333472bf7e
<pre>
Add performJITMemcpyAtomic and simplify jit copying code.
<a href="https://bugs.webkit.org/show_bug.cgi?id=295737">https://bugs.webkit.org/show_bug.cgi?id=295737</a>

Reviewed by Keith Miller.

The eventual goal of this series of patches is to make clear, explicit and separate
entrypoints into code that eventually copies to the jit region or an assembler buffer,
to clearly separate patching that must be atomic and patching that does not.

This is important to ensure that every place relying on atomic copying behaviour
is explicitly documented, so that ARMv7 can ensure the correct alignment of these regions.

To start, we carefuly thread through a new atomic flag, but we avoid changing behavior yet.

The only change this patch should make is that sometimes we can avoid a buffer allocation
when linking.

* Source/JavaScriptCore/assembler/ARM64Assembler.h:
* Source/JavaScriptCore/assembler/ARMv7Assembler.h:
(JSC::ARMv7Assembler::revertJumpTo_movT3movtcmpT2):
(JSC::ARMv7Assembler::revertJumpTo_movT3):
(JSC::ARMv7Assembler::fillNops):
(JSC::ARMv7Assembler::fillNearTailCall):
(JSC::ARMv7Assembler::link):
(JSC::ARMv7Assembler::linkPointer):
(JSC::ARMv7Assembler::relinkCall):
(JSC::ARMv7Assembler::repatchPointer):
(JSC::ARMv7Assembler::replaceWithNops):
(JSC::ARMv7Assembler::setInt32):
(JSC::ARMv7Assembler::setUInt7ForLoad):
(JSC::ARMv7Assembler::setPointer):
(JSC::ARMv7Assembler::linkJumpT1):
(JSC::ARMv7Assembler::linkJumpT2):
(JSC::ARMv7Assembler::linkJumpT3):
(JSC::ARMv7Assembler::linkJumpT4):
(JSC::ARMv7Assembler::linkConditionalJumpT4):
(JSC::ARMv7Assembler::linkBX):
(JSC::ARMv7Assembler::linkConditionalBX):
(JSC::ARMv7Assembler::linkJumpAbsolute):
* Source/JavaScriptCore/assembler/AbstractMacroAssembler.h:
(JSC::AbstractMacroAssembler::emitNops):
* Source/JavaScriptCore/assembler/AssemblerCommon.h:
(JSC::noFlush):
(JSC::memcpyAtomic):
(JSC::memcpyTearing):
(JSC::memcpyAtomicIfPossible):
(JSC::machineCodeCopy):
* Source/JavaScriptCore/assembler/LinkBuffer.cpp:
(JSC::LinkBuffer::copyCompactAndLinkCode):
(JSC::LinkBuffer::linkCode):
* Source/JavaScriptCore/assembler/MacroAssemblerARM64.h:
(JSC::MacroAssemblerARM64::link):
(JSC::MacroAssemblerARM64::reemitInitialMoveWithPatch):
* Source/JavaScriptCore/assembler/MacroAssemblerARMv7.h:
(JSC::MacroAssemblerARMv7::link):
* Source/JavaScriptCore/assembler/RISCV64Assembler.h:
(JSC::RISCV64Assembler::replaceWithNops):
(JSC::RISCV64Assembler::fillNops):
* Source/JavaScriptCore/assembler/X86Assembler.h:
(JSC::X86Assembler::replaceWithHlt):
(JSC::X86Assembler::replaceWithJump):
(JSC::X86Assembler::replaceWithNops):
(JSC::X86Assembler::revertJumpTo_movq_i64r):
(JSC::X86Assembler::revertJumpTo_movl_i32r):
(JSC::X86Assembler::revertJumpTo_cmpl_ir_force32):
(JSC::X86Assembler::revertJumpTo_cmpl_im_force32):
(JSC::X86Assembler::fillNops):
(JSC::X86Assembler::setPointer):
(JSC::X86Assembler::setInt32):
(JSC::X86Assembler::setInt8):
* Source/JavaScriptCore/jit/ExecutableAllocator.cpp:
(JSC::ExecutableMemoryHandle::~ExecutableMemoryHandle):
* Source/JavaScriptCore/jit/ExecutableAllocator.h:
(JSC::jitMemcpyCheckForZeros):
(JSC::jitMemcpyChecks):
(JSC::performJITMemcpy):
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/WTF/wtf/OptionSet.h:
(WTF::ConstexprOptionSet::ConstexprOptionSet):
(WTF::ConstexprOptionSet::operator* const):
* Source/WTF/wtf/StdLibExtras.h:
(WTF::isAligned):
(WTF::sizeof):

Canonical link: <a href="https://commits.webkit.org/298338@main">https://commits.webkit.org/298338@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/aa8cd1b52e9cf582ed63d626e8bac8a0d49b4339

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/114858 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/34599 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/25067 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/121010 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/65580 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/35230 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/43158 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/87308 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42168 "Build was cancelled. Recent messages:Printed configuration; Running apply-patch; Checked out pull request") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/117806 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/28081 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/103145 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67701 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/27259 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/21266 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/64672 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/107190 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/97463 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/21381 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/124207 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/113408 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/41857 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/31282 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/96102 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/42232 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/99336 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95887 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24454 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/41077 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/18917 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/37918 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/41737 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/47255 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/137624 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/41294 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36797 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/44609 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/43037 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/assembler/AssemblerCommon.h:
```diff
@@ -376,31 +378,51 @@ static ALWAYS_INLINE void* memcpyAtomicIfPossible(void* dst, const void* src, si
-static void* performJITMemcpy(void* dst, const void* src, size_t n);
+static ALWAYS_INLINE void* memcpyAtomicIfPossible(void* dst, const void* src, size_t n)
+{
+    if (n == 1 || n == 2 || n == 4 || n == 8)
```
Should we have `n ==16` too?

justinmichaud replied:
In the follow-up, I plan to. I just wanted to make sure this patch didn't change behaviour at first.

Constellation commented on Source/JavaScriptCore/assembler/LinkBuffer.cpp:
```diff
@@ -302,8 +302,10 @@ void LinkBuffer::copyCompactAndLinkCode(MacroAssembler& macroAssembler, JITCompi
 
     uint8_t* codeOutData = m_code.dataLocation<uint8_t*>();
+// It is important not to spill this to the stack.
+#define shouldCopyDirectlyToJITRegion (!m_shouldPerformBranchCompaction || g_jscConfig.useFastJITPermissions)
```
Does it guarantee non-spilling?

justinmichaud replied:
I made this a #define so that we couldn't spill the local. Nothing in this section of code guarantees no spilling, and when I tried to wrap it in a frame-size 0 lambda, it did spill quite a bit already. I just don't want to make it worse

Constellation commented on Source/JavaScriptCore/runtime/OptionsList.h:
```diff
@@ -441,6 +441,7 @@ bool hasCapacityToUseLargeGigacage();
     v(Bool, exitOnResourceExhaustion, false, Normal, nullptr) \
     v(Bool, useExceptionFuzz, false, Normal, nullptr) \
     v(Unsigned, fireExceptionFuzzAt, 0, Normal, nullptr) \
+    v(Bool, fuzzAtomicJITMemcpy, true, Normal, nullptr) \
```
This should be `false`.

justinmichaud replied:
Yes, absolutely! I just wanted EWS to test it first

darinadler commented on Source/WTF/wtf/MathExtras.h:
```diff
@@ -444,6 +444,18 @@ inline constexpr unsigned fastLog2(uint64_t value)
+inline constexpr bool isAligned(T* ptr)
+{
+    return isAligned(ptr, alignment);
+}
```
This doesn’t seem like the right header for this. Pointer things don’t really fall into the "Math" category. There are other suitable headers, like `StdLibExtras.h`. Maybe you can find an even better one.

Constellation commented on Source/JavaScriptCore/assembler/AssemblerCommon.h:
```diff
@@ -26,12 +26,39 @@
+struct RepatchingInfo {
+    const RepatchingAtomic atomic = RepatchingAtomic::Tearing;
+    const RepatchingCopy copy = RepatchingCopy::JITMemcpy;
+    const RepatchingFlush flush = RepatchingFlush::NoFlush;
```
We normally use `{ }` instead of `= ...`

Constellation commented on Source/JavaScriptCore/assembler/AssemblerCommon.h:
```diff
@@ -26,12 +26,39 @@
+enum class RepatchingFlush : uint8_t {
+    Flush,
+    NoFlush,
+};
```
I think we should just use `OptionSet`, like,
```
enum class RepatchingFlag : uint8_t {
    Atomic = 1 << 0,
    JITMemcpy = 1 << 1,
    Flush = 1 << 2,
};
```
And use
```
OptionSet<RepatchingFlag>
```
since `.copy = ...` initialization is not a part of C++.

Constellation reviewed (commented):
Commented.

aoikonomopoulos commented on Source/JavaScriptCore/assembler/ARM64Assembler.h:
```diff
@@ -3659,9 +3662,9 @@ class ARM64Assembler {
+        performJITMemcpy<noFlush(repatch)>(address, buffer, sizeof(int) * NUMBER_OF_ADDRESS_ENCODING_INSTRUCTIONS);
 
-        if (flush)
+        if ((*repatch).contains(RepatchingFlag::Flush))
```
`constexpr`?

aoikonomopoulos commented on Source/JavaScriptCore/assembler/AssemblerCommon.h:
```diff
@@ -379,32 +390,50 @@ static ALWAYS_INLINE void* memcpyAtomicIfPossible(void* dst, const void* src, si
+        // Avoid unaligned accesses.
+        if (WTF::isAligned(dst, n))
+            return memcpyAtomicIfPossible(dst, src, n);
+        return memcpyTearing(dst, src, n);
```
I'm guessing this doesn't look at `RepatchingFlag::Atomic` on 32 bits to reproduce the existing behavior? If so, wouldn't mind a comment explaining that :innocent:

justinmichaud replied:
The follow-up fixes that

aoikonomopoulos commented on Source/JavaScriptCore/assembler/ARMv7Assembler.h:
```diff
@@ -2941,8 +2944,8 @@ class ARMv7Assembler {
-        performJITMemcpy(location - 4, instructions, 4 * sizeof(uint16_t));
-        if (flush)
+        performJITMemcpy<noFlush(repatch)>(location - 4, instructions, 4 * sizeof(uint16_t));
+        if ((*repatch).contains(RepatchingFlag::Flush))
```
`constexpr`?

aoikonomopoulos commented on Source/JavaScriptCore/jit/ExecutableAllocator.cpp:
```diff
@@ -852,9 +852,9 @@ class FixedVMPoolExecutableAllocator final {
             auto emitJumpTo = [&] (void* target) {
                 RELEASE_ASSERT(Assembler::canEmitJump(std::bit_cast<void*>(jumpLocation), target));
                 if (useMemcpy)
```
FWIW it looks like `useMemcpy` is a compile-time decision for all callers of `islandForJumpLocation` too. This is not even a nit, just a note for the future.

kmiller68 commented on Source/WTF/wtf/StdLibExtras.h:
```diff
@@ -159,6 +159,18 @@ inline bool is8ByteAligned(void* p)
+{
+    return isAligned(ptr, alignment);
+}
+
```
This kinda feels like a footgun as a template function. I would expect this to verify my typed pointer is naturally aligned to the expected size for the type.

justinmichaud replied:
I added an assertion

kmiller68 commented on Source/JavaScriptCore/assembler/AssemblerCommon.h:
```diff
@@ -26,12 +26,27 @@
+
+constexpr RepatchingInfo noFlush(RepatchingInfo i)
+{
+    return { *i - *RepatchingInfo { RepatchingFlag::Flush } };
```
Nit: this would be clearer as `return { *i & ~RepatchingFlag::Flush };`

kmiller68 replied:
Oh that's what `-` does on the `OptionSet` nevermind this.

kmiller68 commented on Source/JavaScriptCore/assembler/AssemblerCommon.h:
```diff
@@ -26,12 +26,27 @@
+    Flush = 1 << 2,
+};
+
+using RepatchingInfo = WTF::ConstexprOptionSet<RepatchingFlag>;
```
Why is the ConstexprOptionSet needed? Why can't it just be an OptionSet? Could definitely use a comment in the commit message.

justinmichaud replied:
Happy to comment: A structural type requires all base classes and non-static data members are public and non-mutable

kmiller68 commented on Source/JavaScriptCore/assembler/AssemblerCommon.h:
```diff
@@ -379,32 +390,50 @@ static ALWAYS_INLINE void* memcpyAtomicIfPossible(void* dst, const void* src, si
-static void* performJITMemcpy(void* dst, const void* src, size_t n);
+static ALWAYS_INLINE void* memcpyAtomicIfPossible(void* dst, const void* src, size_t n)
+{
+    if (n == 1 || n == 2 || n == 4 || (!is32Bit() && n == 8))
```
Nit: I would do `isPowerOfTwo(n) && n <= sizeof(CPURegister)`

kmiller68 commented on Source/JavaScriptCore/assembler/AssemblerCommon.h:
```diff
@@ -26,12 +26,27 @@
+
+using RepatchingInfo = WTF::ConstexprOptionSet<RepatchingFlag>;
+
+constexpr RepatchingInfo noFlush(RepatchingInfo i)
```
Nit I would have a `static constexpr RepatchingInfo jitMemcpyRepatch = RepatchingInfo { };` and use that instead of `RepatchingInfo { }` in the template parameters.

kmiller68 reviewed (approved):
r=me

justinmichaud committed:
Add performJITMemcpyAtomic and simplify jit copying code.
https://bugs.webkit.org/show_bug.cgi?id=295737

Reviewed by Keith Miller.

The eventual goal of this series of patches is to make clear, explicit and separate
entrypoints into code that eventually copies to the jit region or an assembler buffer,
to clearly separate patching that must be atomic and patching that does not.

This is important to ensure that every place relying on atomic copying behaviour
is explicitly documented, so that ARMv7 can ensure the correct alignment of these regions.

To start, we carefuly thread through a new atomic flag, but we avoid changing behavior yet.

The only change this patch should make is that sometimes we can avoid a buffer allocation
when linking.

* Source/JavaScriptCore/assembler/ARM64Assembler.h:
* Source/JavaScriptCore/assembler/ARMv7Assembler.h:
(JSC::ARMv7Assembler::revertJumpTo_movT3movtcmpT2):
(JSC::ARMv7Assembler::revertJumpTo_movT3):
(JSC::ARMv7Assembler::fillNops):
(JSC::ARMv7Assembler::fillNearTailCall):
(JSC::ARMv7Assembler::link):
(JSC::ARMv7Assembler::linkPointer):
(JSC::ARMv7Assembler::relinkCall):
(JSC::ARMv7Assembler::repatchPointer):
(JSC::ARMv7Assembler::replaceWithNops):
(JSC::ARMv7Assembler::setInt32):
(JSC::ARMv7Assembler::setUInt7ForLoad):
(JSC::ARMv7Assembler::setPointer):
(JSC::ARMv7Assembler::linkJumpT1):
(JSC::ARMv7Assembler::linkJumpT2):
(JSC::ARMv7Assembler::linkJumpT3):
(JSC::ARMv7Assembler::linkJumpT4):
(JSC::ARMv7Assembler::linkConditionalJumpT4):
(JSC::ARMv7Assembler::linkBX):
(JSC::ARMv7Assembler::linkConditionalBX):
(JSC::ARMv7Assembler::linkJumpAbsolute):
* Source/JavaScriptCore/assembler/AbstractMacroAssembler.h:
(JSC::AbstractMacroAssembler::emitNops):
* Source/JavaScriptCore/assembler/AssemblerCommon.h:
(JSC::noFlush):
(JSC::memcpyAtomic):
(JSC::memcpyTearing):
(JSC::memcpyAtomicIfPossible):
(JSC::machineCodeCopy):
* Source/JavaScriptCore/assembler/LinkBuffer.cpp:
(JSC::LinkBuffer::copyCompactAndLinkCode):
(JSC::LinkBuffer::linkCode):
* Source/JavaScriptCore/assembler/MacroAssemblerARM64.h:
(JSC::MacroAssemblerARM64::link):
(JSC::MacroAssemblerARM64::reemitInitialMoveWithPatch):
* Source/JavaScriptCore/assembler/MacroAssemblerARMv7.h:
(JSC::MacroAssemblerARMv7::link):
* Source/JavaScriptCore/assembler/RISCV64Assembler.h:
(JSC::RISCV64Assembler::replaceWithNops):
(JSC::RISCV64Assembler::fillNops):
* Source/JavaScriptCore/assembler/X86Assembler.h:
(JSC::X86Assembler::replaceWithHlt):
(JSC::X86Assembler::replaceWithJump):
(JSC::X86Assembler::replaceWithNops):
(JSC::X86Assembler::revertJumpTo_movq_i64r):
(JSC::X86Assembler::revertJumpTo_movl_i32r):
(JSC::X86Assembler::revertJumpTo_cmpl_ir_force32):
(JSC::X86Assembler::revertJumpTo_cmpl_im_force32):
(JSC::X86Assembler::fillNops):
(JSC::X86Assembler::setPointer):
(JSC::X86Assembler::setInt32):
(JSC::X86Assembler::setInt8):
* Source/JavaScriptCore/jit/ExecutableAllocator.cpp:
(JSC::ExecutableMemoryHandle::~ExecutableMemoryHandle):
* Source/JavaScriptCore/jit/ExecutableAllocator.h:
(JSC::jitMemcpyCheckForZeros):
(JSC::jitMemcpyChecks):
(JSC::performJITMemcpy):
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/WTF/wtf/OptionSet.h:
(WTF::ConstexprOptionSet::ConstexprOptionSet):
(WTF::ConstexprOptionSet::operator* const):
* Source/WTF/wtf/StdLibExtras.h:
(WTF::isAligned):
(WTF::sizeof):

Canonical link: https://commits.webkit.org/298338@main
